pkg/resources/digitalocean/cloud.go

```diff
@@ -175,3 +183,97 @@ func (c *Cloud) GetApiIngressStatus(cluster *kops.Cluster) ([]kops.ApiIngressSta
 
 	return nil, nil
 }
+
+// FindClusterStatus discovers the status of the cluster, by looking for the tagged etcd volumes
+func (c *Cloud) FindClusterStatus(cluster *kops.Cluster) (*kops.ClusterStatus, error) {
+	etcdStatus, err := findEtcdStatus(c, cluster)
+	if err != nil {
+		return nil, err
+	}
+	status := &kops.ClusterStatus{
+		EtcdClusters: etcdStatus,
+	}
+	klog.V(2).Infof("Cluster status (from cloud): %v", fi.DebugAsJsonString(status))
+	return status, nil
+}
+
+// findEtcdStatus discovers the status of etcd, by looking for the tagged etcd volumes
+func findEtcdStatus(c *Cloud, cluster *kops.Cluster) ([]kops.EtcdClusterStatus, error) {
+	statusMap := make(map[string]*kops.EtcdClusterStatus)
+	volumes, err := getAllVolumesByRegion(c, c.RegionName)
+
+	if err != nil {
+		return nil, fmt.Errorf("error describing volumes: %v", err)
+	}
+
+	for _, volume := range volumes {
+		volumeID := volume.ID
+
+		etcdClusterName := ""
+		var etcdClusterSpec *etcd.EtcdClusterSpec
+
+		for _, myTag := range volume.Tags {
+			klog.V(2).Infof("findEtcdStatus status (from cloud): %v", myTag)
+			// check if volume belongs to this cluster.
+			// tag will be in the format "KubernetesCluster:dev5-k8s-local" (where clusterName is dev5.k8s.local)
+			clusterName := strings.Replace(cluster.Name, ".", "-", -1)
+			if strings.Contains(myTag, fmt.Sprintf("%s:%s", TagKubernetesClusterNamePrefix, clusterName)) {
+				klog.V(10).Infof("findEtcdStatus cluster comparison matched for tag: %v", myTag)
+				// this volume belongs to our cluster, add this to our etcdClusterSpec.
+				// loop through the tags again and
+				for _, volumeTag := range volume.Tags {
+					if strings.Contains(volumeTag, TagKubernetesClusterIndex) {
+						if len(strings.Split(volumeTag, ":")) < 2 {
+							return nil, fmt.Errorf("error splitting the volume tag %q on volume %q", volumeTag, volume)
+						}
+						dropletIndex := strings.Split(volumeTag, ":")[1]
+						etcdClusterSpec, err = c.getEtcdClusterSpec(volume.Name, dropletIndex)
+						klog.V(10).Infof("findEtcdStatus etcdClusterSpec: %v", fi.DebugAsJsonString(etcdClusterSpec))
+						if err != nil {
+							return nil, fmt.Errorf("error parsing etcd cluster tag %q on volume %q: %v", volumeTag, volumeID, err)
+						}
+
+						etcdClusterName = etcdClusterSpec.ClusterKey
+						status := statusMap[etcdClusterName]
+						if status == nil {
+							status = &kops.EtcdClusterStatus{
+								Name: etcdClusterName,
+							}
+							statusMap[etcdClusterName] = status
+						}
+
+						memberName := etcdClusterSpec.NodeName
+						status.Members = append(status.Members, &kops.EtcdMemberStatus{
+							Name:     memberName,
+							VolumeId: volume.ID,
+						})
+					}
+				}
+			}
+		}
+	}
+
+	var status []kops.EtcdClusterStatus
```

Comment:
This can be improved slightly performance-wise since we know the target size of the slice:

```go
status := make([]kops.EtcdClusterStatus, 0, len(statusMap))
```

Reply:
sure, updated.